src/audio/mixin_mixout/mixin_mixout_generic.c

```diff
@@ -0,0 +1,403 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2022 Intel Corporation. All rights reserved.
+//
+// Author: Andrula Song <xiaoyuan.song@intel.com>
+
+#include <ipc4/mixin_mixout.h>
+#include <sof/common.h>
+
+#ifdef MIXIN_MIXOUT_GENERIC
+
+#if CONFIG_FORMAT_S16LE
+/* Instead of using sink->channels and source->channels, sink_channel_count and
+ * source_channel_count are supplied as parameters. This is done to reuse the function
+ * to also mix an entire stream. In this case the function is called with fake stream
+ * parameters: multichannel stream is treated as single channel and so the entire stream
+ * contents is mixed.
+ */
+static void normal_mix_channel_s16(struct audio_stream __sparse_cache *sink, uint32_t start_frame,
+				   uint32_t mixed_frames,
+				   const struct audio_stream __sparse_cache *source,
+				   uint32_t frame_count, uint16_t gain)
+{
+	uint32_t frames_to_mix, frames_to_copy, left_frames;
+	uint32_t n, nmax, i;
+
+	/* audio_stream_wrap() is required and is done below in a loop */
+	int16_t *dst = (int16_t *)sink->w_ptr + start_frame;
+	int16_t *src = (int16_t *)source->r_ptr;
+
+	assert(mixed_frames >= start_frame);
+	frames_to_mix = MIN(mixed_frames - start_frame, frame_count);
+	frames_to_copy = frame_count - frames_to_mix;
+
+	for (left_frames = frames_to_mix; left_frames > 0; left_frames -= n) {
+		src = audio_stream_wrap(source, src);
+		dst = audio_stream_wrap(sink, dst);
+		/* calculate the remaining samples*/
+		nmax = audio_stream_samples_without_wrap_s16(source, src);
+		n = MIN(left_frames, nmax);
+		nmax = audio_stream_samples_without_wrap_s16(sink, dst);
+		n = MIN(n, nmax);
+		for (i = 0; i < n; i++) {
+			*dst = sat_int16((int32_t)*dst + (int32_t)*src++);
```

Comment:
for (i = 0; i < n; i++) 
			*dst++ = sat_int16((int32_t)*dst + (int32_t)*src++);

looks better?

Reply:
There will be compilation errors when writing like this

Reply:
I don't see why these `(uint32_t)` casts are needed. Casts hide bugs, especially when dealing with pointers and pointer arithmetic. Remove these casts or add a comment explaining why they are required.

Reply:
Unsigned overflow is defined in the standard (and the arguments may even be promoted to `int` before the addition takes, can't remember the details).

Reply:
> *dst++ = sat_int16((int32_t)*dst + (int32_t)*src++);

Wow, just noticed this. This is incredibly dangerous because you cannot tell whether the `dst` on the right side will be incremented by the left side before or after the right side is evaluated.

> There will be compilation errors when writing like this

Which one?

Reply:
>you are right , if make it like *dst++ = sat_int16((int32_t)*dst + (int32_t)*src++); would lesd to compilation errors

Reply:
Don't "resolve" and hide unfinished conversations.

It's not required to resolve all conversations.

Reply:
Sorry, since I remove that two (uint32_t), I thought this has done.

Reply:
Some threads have more than one topic.